frontend/packages/console-dynamic-plugin-sdk/src/utils/fetch/co-fetch.ts

```diff
@@ -0,0 +1,141 @@
+
+  const fetchPromise = fetch(url, allOptions).then((response) =>
+    validateStatus(response, url, allOptions.method, retry),
+  );
+
+  // return fetch promise directly if timeout <= 0
+  if (timeout < 1) {
+    return fetchPromise;
+  }
+
+  const timeoutPromise: Promise<Response> = new Promise((unused, reject) =>
+    setTimeout(() => reject(new TimeoutError(url, timeout)), timeout),
+  );
+
+  // Initiate both the fetch promise and a timeout promise
+  return Promise.race([fetchPromise, timeoutPromise]);
+};
+
+export const coFetch: CoFetch = async (
+  url: string,
+  options: RequestInit = {},
+  timeout: number = 60000,
+): Promise<Response> => {
+  let attempt = 0;
+  let response;
+  let retry = true;
+  while (retry) {
+    retry = false;
+    attempt++;
+    try {
+      // have to disable for retry logic
+      // eslint-disable-next-line no-await-in-loop
+      response = await coFetchInternal(url, options, timeout, attempt < 3);
+    } catch (e) {
+      if (e instanceof RetryError) {
+        retry = true;
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  return response;
+};
+
+const coFetchCommon = async (
+  url: string,
+  method: string = 'GET',
+  options: RequestInit = {},
+  timeout?: number,
+) => {
+  const headers = getImpersonateHeaders() || {};
+  // Pass headers last to let callers to override Accept.
+  const allOptions = _.defaultsDeep({ method }, options, { headers });
+  const response = await coFetch(url, allOptions, timeout);
+  const text = await response.text();
+  const isPlainText = response.headers.get('Content-Type') === 'text/plain';
+  if (!text) {
+    return isPlainText ? '' : {};
+  }
+  return isPlainText || !response.ok ? text : JSON.parse(text);
+};
+
+export const coFetchJSON: CoFetchJSON = (
+  url: string,
+  method: string = 'GET',
+  options: RequestInit = {},
+  timeout?: number,
+) => {
+  const allOptions = _.defaultsDeep({}, options, { headers: { Accept: 'application/json' } });
+  return coFetchCommon(url, method, allOptions, timeout);
+};
+
+export const coFetchText: CoFetchText = (
+  url: string,
+  options: RequestInit = {},
+  timeout?: number,
+): Promise<string> => {
+  return coFetchCommon(url, 'GET', options, timeout);
+};
+
+const coFetchSendJSON = (
+  url: string,
+  method: string,
+  json = null,
+  options: RequestInit = {},
+  timeout: number,
+) => {
+  const allOptions: Record<string, any> = {
+    headers: {
+      Accept: 'application/json',
+      'Content-Type': `application/${
+        method === 'PATCH' ? 'json-patch+json' : 'json'
+      };charset=UTF-8`,
+    },
+  };
+  if (json) {
+    allOptions.body = JSON.stringify(json);
+  }
+  return coFetchJSON(url, method, _.defaultsDeep(allOptions, options), timeout);
+};
+
+coFetchJSON.delete = (url: string, json = null, options: RequestInit = {}, timeout?: number) => {
+  return json
+    ? coFetchSendJSON(url, 'DELETE', json, options, timeout)
+    : coFetchJSON(url, 'DELETE', options, timeout);
+};
```

Comment:
```suggestion
coFetchJSON.delete = (url, json = null, options = {}, timeout) => {
  return json
    ? coFetchSendJSON(url, 'DELETE', json, options, timeout)
    : coFetchJSON(url, 'DELETE', options, timeout);
};
```

+ same for the rest `post`, `put`, `patch`

Reply:
Done.